[Example]: Multimodal RAG with LanceDB

AyushExel opened the pull request:
Multimodal RAG with LanceDB.

What this does:
- Fetches a product catalog from a live API (fakestoreapi.com).
- Embeds product descriptions using a CLIP model (clip-ViT-B-32) and stores them in LanceDB.
- Implements a RAG agent with a `find_products` tool that can:
    1. Perform semantic search (e.g., "something for the cold weather").
    2. Perform metadata filtering (e.g., "show me all electronics").
       (e.g., "a cool t-shirt" from the "men's clothing" category under $20).
- Generates and displays an image collage of the results for instant visual feedback.
- Creates logfire tracing dashboard if api key is set

Install dependencies:
```
    pip install lancedb sentence-transformers torch httpx pandas Pillow logfire[httpx]
```

Set your Google API key (for the agent's text generation):
```
    export GOOGLE_API_KEY=your_api_key_here
```

Usage:
    First, build the product database from the live API
```
    python lancedb_multimodal.py build
```

    # Then, ask for a recommendation:
```
    python lancedb_multimodal.py search "a cool t-shirt in men's clothing under 20 dollars"
```

Kludex commented on pyproject.toml:
```diff
@@ -47,13 +47,14 @@ requires-python = ">=3.9"
 examples = ["pydantic-ai-examples=={{ version }}"]
 logfire = ["logfire>=3.11.0"]
 a2a = ["fasta2a>=0.4.1"]
+lancedb = ["pydantic-ai-slim[lancedb]"]
```
This needs to be removed. We don't want to add `lancedb` as an extra dependency of `pydantic-ai`.

Kludex commented on pyproject.toml:
```diff
@@ -47,13 +47,14 @@ requires-python = ">=3.9"
 [tool.hatch.metadata.hooks.uv-dynamic-versioning]
 dependencies = [
-    "pydantic-ai-slim[openai,vertexai,google,groq,anthropic,mistral,cohere,bedrock,huggingface,cli,mcp,evals,ag-ui,retries]=={{ version }}",
+    "pydantic-ai-slim[openai,vertexai,google,groq,anthropic,mistral,cohere,bedrock,huggingface,cli,mcp,evals,ag-ui,retries,lancedb]=={{ version }}",
```
Revert please.

Kludex commented on pydantic_ai_slim/pyproject.toml:
```diff
@@ -74,6 +74,7 @@ huggingface = ["huggingface-hub[inference]>=0.33.5"]
 # Tools
 duckduckgo = ["ddgs>=9.0.0"]
 tavily = ["tavily-python>=0.5.0"]
+lancedb = ["lancedb>=0.6.1"]
```
Revert please.

Kludex commented:
> I need some help with managing deps. It seems like lancedb is not being detected during test time even though I've added it to examples deps list.

You need to add the dependency you need for the docs on the `dev = ` in the root `pyproject.toml`.

Kludex commented on examples/pydantic_ai_examples/lancedb_multimodal.py:
```diff
@@ -0,0 +1,277 @@
+
+
+# ruff: noqa
+# pyright: reportMissingImports=false
+# pyright: reportInvalidTypeForm=false
+# pyright: reportUnknownVariableType=false
+# pyright: reportUnknownMemberType=false
+# pyright: reportUnknownArgumentType=false
+# pyright: reportUntypedBaseClass=false
```
Why so many?

samuelcolvin replied:
to be clear, we're happy to accept an example from you, but this is just a big middle finger to how we operate.

it's either saying:
* I don't care about any of your typing, linting, DX work
* or, we're special, we don't need to do this

Either lanceDB is typesafe, in which case remove these. or LanceDB is not type-safe in which case the example probably doesn't belong in our docs!

AyushExel replied:
Added these to figure out the deps part while fighting CI.. Will revert back

Kludex commented on examples/pydantic_ai_examples/lancedb_multimodal.py:
```diff
@@ -0,0 +1,277 @@
+    from lancedb.pydantic import LanceModel, Vector
+    from PIL import Image
+    from sentence_transformers import SentenceTransformer
+except ImportError:
+    print(
+        """Missing dependencies. To run this example, please install the required packages by running:
+pip install lancedb sentence-transformers torch httpx pandas Pillow logfire[httpx]"""
+    )
+    sys.exit(0)
```
Please follow what we do in `weather_agent_gradio.py`.

samuelcolvin commented on examples/pyproject.toml:
```diff
@@ -63,6 +63,8 @@ dependencies = [
     "duckdb>=1.3.2",
     "datasets>=4.0.0",
     "pandas>=2.2.3",
+    "lancedb",
+    "sentence-transformers"
```
please include a version bound.

samuelcolvin commented on examples/pydantic_ai_examples/lancedb_multimodal.py:
```diff
@@ -0,0 +1,277 @@
+"""Multimodal RAG with LanceDB.
```
this should go in the markdown, not in code.

AyushExel commented:
Addressed specific feedbacks:
1. Moved lancedb out of separate dependency group to dev
2. Removed hotfixes for pyright  - except a couple of places involving dynamic types, similar to https://github.com/pydantic/pydantic-ai/blob/main/examples/pydantic_ai_examples/rag.py#L189 . Not sure if there's a better practice. Happy to be educated.
3. Removed top-level docstring guide from code file - Might want to double check this, as the other examples have them.
